### ISSUE TITLE
perplexity computation for self debiasing

### ISSUE THREAD

**Issue by boxin-wbx** [timoschick/self-debiasing#2]
Hi,

Thanks for open-sourcing the code!

I found that [line 220-239 in modeling.py](https://github.com/timoschick/self-debiasing/blob/762f1811cc361014e92f0b6f7ffdd7f3e8d3200d/modeling.py#L220)  a little bit confusing. Specifically, I have the following questions:

1. Why do we need to flip the attention mask for input prefixes (`input_prefixes['attention_mask'] = torch.flip(input_prefixes['attention_mask'], dims=[1])`)?
2. Why do we need to roll the `input_prefixes['input_ids']` by the length of input_prefixes? 
3. From my understanding, we can simply concat the `input_prefixes` without padding to `input_ids_repeated` (same for attention mask) and it is done.
4. Why do we need to use `shifts[0]`? Isn't `shifts[0]` always 0 because the first prefix is `['']`?

Thanks in advance!

**Comment by timoschick** [comment]
Lines 220-239 make sure that the `input_prefixes` are padded on the **left** before concatenating them to the `input_ids_repeated`. Let's say our actual input is `"x y z"` and our `input_prefixes` are `""`, `"a"` and `"a b c"`. Then the encoded versions would be
```
>>> input_prefixes
[[PAD, PAD, PAD],
 ["a", PAD, PAD],
 ["a", "b", "c"]]

>>> input_ids_repeated
[["x", "y", "z"],
 ["x", "y", "z"],
 ["x", "y", "z"]]
```
If we do what you propose in (3), we get:
```
>>> torch.cat([input_prefixes, input_ids_repeated], dim=1)
[[PAD, PAD, PAD, "x", "y", "z"],
 ["a", PAD, PAD, "x", "y", "z"],
 ["a", "b", "c", "x", "y", "z"]]
```
Lines 220-239 instead create the following output (and corresponding attention mask) with no **intermediate** padding tokens:
```
[[PAD, PAD, PAD, "x", "y", "z"],
 [PAD, PAD, "a", "x", "y", "z"],
 ["a", "b", "c", "x", "y", "z"]]
```
To answer your questions in more detail:
> 1. Why do we need to flip the attention mask for input prefixes (input_prefixes['attention_mask'] = torch.flip(input_prefixes['attention_mask'], dims=[1]))?

This makes sure that the padding to the left is properly reflected in the attention mask.

> 2. Why do we need to roll the input_prefixes['input_ids'] by the length of input_prefixes?

Similarly, this makes sure that the `input_prefixes` are padded to the left.

> 3. From my understanding, we can simply concat the input_prefixes without padding to input_ids_repeated (same for attention mask) and it is done.

See above.

> 4. Why do we need to use shifts[0]? Isn't shifts[0] always 0 because the first prefix is ['']?

No, `shifts` is computed as 
```
shifts = input_prefixes['attention_mask'].shape[-1] - input_prefixes['attention_mask'].sum(dim=-1)
```
In the above example, this would be equal to
```
shifts = 3 - torch.tensor([0, 1, 3]) = torch.tensor([3, 2, 0])
```
So `shifts[0] = 3`.

**Comment by boxin-wbx** [comment]
Thanks so much for your explanation. I got you. The idea of adding PAD to the left is amazing!
